### ISSUE TITLE
Could the code within this if/else statement be written more efficiently?

### ISSUE THREAD

**Issue by hdsbbrooks** [HDSB-eLearning-ICS4UO-202021/Class-Examples#1]
This section of code within the if/else statement could be written in different ways.  Rewrite this code in another way to achieve the same functionality.  Do you think your version is more/less/equivalent efficiency to the original code?  Explain why you think this.

https://github.com/HDSB-eLearning-ICS4UO-202021/Class-Examples/blob/6278f35bf97992aff53c3bbca6cb7f46c062f45d/Intro%20to%20OO/Game%20of%20Life/Cell.java#L86-L106

**Comment by Dani3lx** [comment]
Here is the new code

       if (living) {

            if ( numberNeighboursLiving == 2 || numberNeighboursLiving == 3 ) {
                return true;
            }
            else {
                return false;
            }
            
        }
        else {
             if (numberNeighboursLiving == 3) {
                 return true;
             }
        }

This code is more efficient as it eliminates the need for else-if statement.

**Comment by nachoroo** [comment]
If you really wanted to save lines, you could simplify the if-else statement and  rewrite the code with the ternary operator:
```
if (living) {
    return ( numberNeighboursLiving >= 2 && numberNeighboursLiving <= 3 ) ? true : false;  
} else {
    if (numberNeighboursLiving == 3) {
        return true;
    }
}
```
In effect, this code would be slightly more efficient, as it combines both survival criteria into one expression and reduces the nested if-then-else statement. The code is also a bit more compact, however, it is a little harder to read. If you were looking to write as few lines as possible, you could reduce it even further to this:
```
return living ? (numberNeighboursLiving >= 2 && numberNeighboursLiving <= 3) ? true : false : (numberNeighboursLiving == 3) ? true : false;
```
but then this becomes very impractical, messy, and hard to understand.

**Comment by bepapab** [comment]
```
if (living) {
           switch (numberNeighboursLiving) {
                case 2: return true;
                case 3: return true;
                default: return false; 
            }
        }
        else { //Cell is dead
             if (numberNeighboursLiving == 3) {
                 return true;
             }
        }
```

This code is easier to read and more effecient because it only does 2 comparisons.

**Comment by cz-code** [comment]
One way to make make this section of code more efficient is by utilizing a switch statement. The switch statement will reduce the number of lines needed and improve readability since compares the number of living neighbours (numberNeighboursLiving) to each case. Only one switch statement is required opposed using 3 blocks of code for "if", "else if", and "else".

```
if (living) { //cell is alive
  switch (numberNeighboursLiving) {
    case 2:
    case 3: return true; //cell will live if it has two or three live neighbours
    default: return false; //cell will die when number of live neighbour cells is not equal to two or three
  }
  
} else { //cell is dead
  //dead cell with three neighbours becomes live cell.
  if (numberNeighboursLiving == 3) {
    return true;
  }
}
```

**Comment by ParmesanCheese345** [comment]
```Java
 if (living) {
        if (numberNeighboursLiving == 2 || numberNeighboursLiving == 3 ) 
            return true;
        else 
            return false;
    }
    else {
         if (numberNeighboursLiving == 3) 
             return true;
    }
 ```
My code is like previously posted solutions. All that is required is to merge the if and else-if statements into one if statement. This code can be written more quickly and is slightly more efficient. However, the increased efficiency will only be noticeable when repeating the code hundreds of thousands of times.

**Comment by GalaxyBoy1234** [comment]
[Effiency.txt](https://github.com/HDSB-eLearning-ICS4UO-202021/Class-Examples/files/6069054/Effiency.txt)

The code is easier to read and is more understandable

**Comment by jules-77** [comment]
I combined the if and else if lines so that the result of them being true, returns true. This way it seems like there are less events that could occur. The changes would make the program more efficient because there is not an else-if statement. With that being said, that is such a small change it would be essentially unnoticeable.
[Correction.txt](https://github.com/HDSB-eLearning-ICS4UO-202021/Class-Examples/files/6069411/Correction.txt)

**Comment by 1ligra** [comment]
```
 if (living) { 
	return (numberNeighoursLiving == 2 || numberNeighoursLiving == 3) ? true : false;
} else {
	if (numberNeighoursLiving == 3) {
		return true;
	}
}
```
	
I basically just condensed the first part using operators.

**Comment by Cheeseburger1776** [comment]
I agree with all the above answers. The code becomes more concise and efficient. Here is another option; 

It keeps the code easy to read, but not less efficient. Anyways, its an option for the programmer.

[Correction.txt](https://github.com/HDSB-eLearning-ICS4UO-202021/Class-Examples/files/6080724/Correction.txt)

**Comment by VernierLabQuest** [comment]
I had the same answer basic answer as several people above, in simply combining the inner if statements. 
``` Java
        if (living) {
            if ( numberNeighboursLiving < 2||numberNeighboursLiving > 3 ) {
                return false;
            }
            //* Any live cell with two or three live neighbours lives on to the next generation.
            else if ( numberNeighboursLiving <= 3 ) {
                return true;
            }
             //* Any live cell with more than three live neighbours dies, as if by overpopulation.
        }
        else { //Cell is dead
             //* Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
             if (numberNeighboursLiving == 3) {
                 return true;
             }
        }
```
Alternatively, a switch statement could also work inside the if statement. 
``` Java
       if (living) {
            //* Any live cell with fewer than two live neighbours dies, as if caused by underpopulation.
            switch(numberNeighboursLiving){
                case 2:
                case 3:
                return true;
                default:
                return false;
            }
        }
        else { //Cell is dead
             //* Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
             if (numberNeighboursLiving == 3) {
                 return true;
             }
        }
```
Either way, the code runs the same with slight speed improvements.
